src/globus_cli/login_manager/auth_flows.py

```diff
@@ -87,7 +89,25 @@ def exchange_code_and_store(auth_client, auth_code):
     """
     Finishes auth flow after code is gotten from command line or local server.
     Exchanges code for tokens and stores them.
+
+    A user may have a different identity this time than what they previously logged in
+    as, so to secure incremental auth flows, if the new tokens don't match the previous
+    identity we revoke them and instruct the user to logout before continuing.
     """
     adapter = token_storage_adapter()
     tkn = auth_client.oauth2_exchange_code_for_tokens(auth_code)
+    sub_new = tkn.decode_id_token()["sub"]
+    auth_user_data = adapter.read_config(_STORE_CONFIG_SUB_NAME)
+    if auth_user_data and sub_new != auth_user_data.get("sub"):
+        auth_client.oauth2_revoke_token(tkn["access_token"])
+        auth_client.oauth2_revoke_token(tkn["refresh_token"])
```

Comment:
The token response will contain multiple new tokens, so we should make sure to revoke them all in this step.

I'm not sure what we should do if revocation fails (e.g. network goes down at this exact moment).
I think maybe -- and I'm not sure about this -- we want to wrap this in a try except block and print a warning if that happens?

Reply:
Thanks. For the first part, `.by_resource_server` is a comprehensive enumeration right?
```python
for rs, tokens in tkn.by_resource_server.items():
    auth_client.oauth2_revoke_token(tokens["access_token"])
    auth_client.oauth2_revoke_token(tokens["refresh_token"])
```

> I'm not sure what we should do if revocation fails
I wouldn't be opposed to a try/except to catch such errors, log it, and just ask the user to try again. Do you think that sort of thing makes sense within `.oauth2_revoke_token` or at call sites like here?

Reply:
Yep, `by_resource_server` will get them all. 👍

> ... just ask the user to try again ...

The trouble is that the error is happening while we're handling a usage error to begin with.
So we don't really want them to try the exact same operation again -- and even if they did, we'd get new and different tokens back.

We *could* put another blob into the token storage DB, but that's dedicating much more effort to this scenario than I think is appropriate.

It's not necessarily a problem if the CLI hard-errors in this case on failing revocation. The user would get a slightly surprising error message, but maybe that's acceptable.

Perhaps the best course of action is to use a `try-finally` construct (no `except` block) and put the error message echo in the `finally`. That way, we're guaranteed to show our desired message, even if the revocation fails. And revocation failure would show the underlying error.

Reply:
Looking at the `try-finally`, I know it's weird but I _think_ it's the right thing to do.
We can definitely revisit if we have some clear idea about improved handling in the future.

I'm going to mark this resolved.